mcneilco/acas
# Fix logout, redirect issue and saml syncing issue

## Description
### The bug
**Steps to Reproduce:** 
Click logout
**Expected Results:** 
User is logged out and redirected to the homepage
**Actual Results:** 
Clicking logout results in this error being displayed on a page to the user
```
Error: req#logout requires a callback function
--
at IncomingMessage.req.logout.req.logOut (/home/runner/build/node_modules/passport/lib/http/request.js:65:44)
at exports.logout (/home/runner/build/routes/loginRoutes.js:111:7)
at Layer.handle [as handle_request] (/home/runner/build/node_modules/express/lib/router/layer.js:95:5)
at next (/home/runner/build/node_modules/express/lib/router/route.js:144:13)
at Route.dispatch (/home/runner/build/node_modules/express/lib/router/route.js:114:3)
at Layer.handle [as handle_request] (/home/runner/build/node_modules/express/lib/router/layer.js:95:5)
at /home/runner/build/node_modules/express/lib/router/index.js:284:15
at param (/home/runner/build/node_modules/express/lib/router/index.js:365:14)
at param (/home/runner/build/node_modules/express/lib/router/index.js:376:14)
at Function.process_params (/home/runner/build/node_modules/express/lib/router/index.js:421:3)
```

**Extra Information:**
It look like our package.json had a mistaken setting of ">=0.4.1" which means we pick up any changes to the passport package.
In this case, we picked up version 0.6.0 which, among a bunch of security fixes, also made logout asynchronous.
Stack over flow of same issue: https://stackoverflow.com/questions/72336177/error-reqlogout-requires-a-callback-function
https://medium.com/passportjs/fixing-session-fixation-b2b68619c51d

Redirects after login also seem to be broken. i.e.
if I request http://localhost:3000/cmpdreg 
I am redirected to http://localhost:3000/login?redirect_url=http://localhost:3000/cmpdreg
After login I am not redirected to http://localhost:3000/cmpdreg

### The Fix

 - Lock package.json to current versions rather than `>` so we don't inadvertently pick up more changes
 - Fix logout by passing the next steps as a callback to req.logout
 - Fix redirects by adding `keepSessionInfo: true` to authentication
 - In the course of fixing this, we also found that saml auth was broken when sync users was true
   - Fix was to add back authorRoutes.syncUsers function[https://github.com/mcneilco/acas/pull/946](https://github.com/mcneilco/acas/pull/946)


## Related Issue
ACAS-341
ACAS-342

## How Has This Been Tested?
Ran manual tests for
 - Logout when using direct login method (not saml)
 - Login when using direct login method
 - Login when using saml authentication and sync roles true
 - When logged out, requesting an experiment path and a cmpd reg path and verifying I was properly redirected:
    - e.g. requested http://localhost:3000/experiment_base/codeName/EXPT-00000002
    - logged in
    - was properly redirected back to: http://localhost:3000/experiment_base/codeName/EXPT-00000002 after login
